src/htm/encoders/SimHashDocumentEncoder.cpp

```diff
@@ -0,0 +1,253 @@
+/* -----------------------------------------------------------------------------
+ * HTM Community Edition of NuPIC
+ * Copyright (C) 2016, Numenta, Inc. https://numenta.com
+ *               2019, David McDougall
+ *               2019, Brev Patterson, Lux Rota LLC, https://luxrota.com
+ *
+ * This program is free software: you can redistribute it and/or modify it
+ * under the terms of the GNU Affero Public License version 3 as published by
+ * the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE. See the GNU Affero Public License for
+ * more details.
+ *
+ * You should have received a copy of the GNU Affero Public License along with
+ * this program.  If not, see http://www.gnu.org/licenses.
+ * -------------------------------------------------------------------------- */
+
+/** @file
+ * SimHashDocumentEncoder.cpp
+ */
+
+#include <algorithm>  // transform
+#include <bitset>     // to_string
+#include <cctype>     // tolower
+#include <climits>    // CHAR_BIT
+#include <regex>
+
+#include <hasher.hpp> // digestpp: sha3+shake256 hash digests
+#include <algorithm/sha3.hpp>
+#include <algorithm/shake.hpp>
+
+#include <htm/encoders/SimHashDocumentEncoder.hpp>
+
+
+namespace htm {
+
+  /**
+   * Constructor
+   * @see SimHashDocumentEncoder.hpp
+   */
+  SimHashDocumentEncoder::SimHashDocumentEncoder(const SimHashDocumentEncoderParameters &parameters)
+  {
+    initialize(parameters);
+  }
+
+  /**
+   * Initialize
+   * @see SimHashDocumentEncoder.hpp
+   */
+  void SimHashDocumentEncoder::initialize(const SimHashDocumentEncoderParameters &parameters)
+  {
+    // Pre-processing, Check parameters
+    NTA_CHECK((parameters.activeBits > 0u) || (parameters.sparsity > 0.0f))
+      << "Need only one argument of: 'activeBits' or 'sparsity'.";
```

Comment:
this check needs be improved. Does not do what the comment says. 
- you could use XOR here
- or check other ways: 
  - activeBits or sparsity (any of) AND the exactly one must be ON

Reply:
woo, thanks.  will do.